Allow to override toolchain from rust-toolchain file

thomaseizinger opened the pull request:
Resolves #21. 

I tried to test the new feature as much as possible. When adding more tests, I had to introduce `mock-env` to make them consistently pass. I guess this is because they are run in parallel.

To actually test the override feature, I had to make the path to the file configurable. Not really pretty but I thought it is better than not having it tested at all.

The code also assumes that it is run from the cloned repository as the current working directory but that is kind of a given anyway if you want to use that feature. If the file doesn't exist, we try to parse the toolchain from the input variable as it is currently done.

svartalf commented on src/args.ts:
```diff
@@ -10,17 +11,30 @@ export interface ToolchainOptions {
+        return readFileSync(overrideFile, {
+            encoding: "utf-8",
+            flag: "r"
+        })
```
Can you add `String.trim()` here? I'm not sure how exactly `rustup` invocation later will behave if one of arguments will contain a line break or other whitespace character.

thomaseizinger replied:
Ah good catch! I was actually thinking of that but then somehow forgot about it. Will add some tests for this as-well :)

thomaseizinger replied:
Ah good catch! I was actually thinking of that but then somehow forgot about it. Will add some tests for this as-well :)

thomaseizinger replied:
Please see 0cea29e :)

thomaseizinger commented:
@svartalf What are your plans on merging this?

svartalf commented:
@thomaseizinger I'm sorry for being so long with that PR.

So, I started testing it and I'm probably missing something here, but with `rust-toolchain` file available it fails to find it properly; check this out:

1. https://github.com/actions-rs/example/commit/fc798f969ee01d1e5e9cef9cd60b0b0d98927cef is using Action input as usual, CI is [green](https://github.com/actions-rs/example/commit/fc798f969ee01d1e5e9cef9cd60b0b0d98927cef/checks?check_suite_id=393082346)
2. https://github.com/actions-rs/example/commit/29442b950d15fa303fbaf26b592304f60a2fcc22 is using `rust-toolchain` file with `toolchain` input removed, CI [is failing](https://github.com/actions-rs/example/commit/29442b950d15fa303fbaf26b592304f60a2fcc22/checks?check_suite_id=393087284) to determine toolchain version
3. https://github.com/actions-rs/example/commit/a7be715bf4550f00b62a34490f798f85a5991292 is using both `toolchain` input and `rust-toolchain` file; [CI is using input](https://github.com/actions-rs/example/commit/a7be715bf4550f00b62a34490f798f85a5991292/checks?check_suite_id=393092778) instead of file

My guess is that CWD in this line `path.join(process.cwd(), "rust-toolchain");` is probably pointing to the different directory? Do you have any thoughts on that?

thomaseizinger commented:
@svartalf 

> My guess is that CWD in this line `path.join(process.cwd(), "rust-toolchain");` is probably pointing to the different directory? Do you have any thoughts on that?

That is interesting. If the current working directory is _not_ the repository folder, how do any of the other actions work? :D

I'll do some more research on this.

thomaseizinger commented:
@svartalf I found the problem..

I didn't run `npm run build` as part of the commit, which means all the tests in CI didn't work and also the commit you referenced was actually using the old code.

I committed the output of `npm run build` now and everything works :D

To prove that, I also added a CI step and enabled the CI for pull-requests.

I tried to write a script for CI that checks if the `dist` folder was modified if typescript files were modified but it was too brittle...
It is a really annoying aspect of Github actions that you have to keep the build result up to date as part of your source tree.

What is your local development workflow? Do you have local commit hooks that prevent it from happening that you forget to run `npm run build`?

thomaseizinger commented:
> What is your local development workflow? Do you have local commit hooks that prevent it from happening that you forget to run npm run build?

I guess the new CI step actually guarantees this, at least for this feature. It is kind of a shame though that the unit tests are not enough :/

thomaseizinger commented:
I've added a pre-commit hook that should catch these errors early on. To automatically "ship" this pre-commit hook to every developer, I've added a `preTest` script to the package.json file which sets the git hooks path to the folder set in the repository.

This is neat but can potentially have one problem: if some developer wants to have their on git hooks _locally_ for this repository, then we break them.
I'll leave it up to you what is more important:

- making sure the check is run for every developer
- respect the local git hooks of every developer

If you disagree, simply drop the last commit of this PR :)
